dm/dm/master/scheduler/scheduler_test.go

```diff
@@ -1967,3 +1974,71 @@ func (t *testScheduler) TestUpgradeCauseConflictRelayType(c *C) {
 	c.Assert(worker.RelaySourceID(), HasLen, 0)
 	c.Assert(s.workers[workerName2].Stage(), Equals, WorkerFree)
 }
+
+func (t *testScheduler) TestOperateValidatorTask(c *C) {
+	defer clearTestInfoOperation(c)
+
+	var (
+		logger      = log.L()
+		s           = NewScheduler(&logger, config.Security{})
+		sourceID1   = "mysql-replica-1"
+		workerName1 = "dm-worker-1"
+		taskName    = "task-1"
+		keepAlive   = int64(2)
+		subtaskCfg  config.SubTaskConfig
+	)
+	c.Assert(subtaskCfg.DecodeFile(subTaskSampleFile, true), IsNil)
+	subtaskCfg.SourceID = sourceID1
+	subtaskCfg.Name = taskName
+	subtaskCfg.ValidatorCfg = config.ValidatorConfig{Mode: config.ValidationNone}
+	c.Assert(subtaskCfg.Adjust(true), IsNil)
+
+	workerInfo1 := ha.WorkerInfo{Name: workerName1}
+	bound := ha.SourceBound{
+		Source: sourceID1,
+		Worker: workerName1,
+	}
+	sourceCfg, err := config.LoadFromFile("../source.yaml")
+	c.Assert(err, IsNil)
+	s.etcdCli = etcdTestCli
+	sourceCfg.SourceID = sourceID1
+	_, err = ha.PutSourceCfg(etcdTestCli, sourceCfg)
+	c.Assert(err, IsNil)
+	_, err = ha.PutWorkerInfo(etcdTestCli, workerInfo1)
+	c.Assert(err, IsNil)
+	_, err = ha.PutSourceBound(etcdTestCli, bound)
+	c.Assert(err, IsNil)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	//nolint:errcheck
+	go ha.KeepAlive(ctx, etcdTestCli, workerName1, keepAlive)
+	c.Assert(s.recoverSources(), IsNil)
+	_, err = s.recoverWorkersBounds()
+	c.Assert(err, IsNil)
+	c.Assert(s.Start(ctx, etcdTestCli), IsNil)
+	// CASE 1: start subtask without starting validation
+	c.Assert(s.AddSubTasks(false, subtaskCfg), IsNil) // create new subtask without validation
+	t.subTaskCfgExist(c, s, subtaskCfg)
+	subtaskCfg.ValidatorCfg.Mode = config.ValidationFull // set mode
+	stCfgs := make(map[string]map[string]config.SubTaskConfig)
+	stCfgs[subtaskCfg.Name] = make(map[string]config.SubTaskConfig)
+	stCfgs[subtaskCfg.Name][subtaskCfg.SourceID] = subtaskCfg
+	c.Assert(s.OperateValidationTask(pb.Stage_Running, stCfgs), IsNil)                      // create validator task
+	t.validatorStageMatch(c, s, subtaskCfg.Name, subtaskCfg.SourceID, pb.Stage_Running)     // task running
+	t.validatorModeMatch(c, s, subtaskCfg.Name, subtaskCfg.SourceID, config.ValidationFull) // succeed to change mode
+	c.Assert(s.OperateValidationTask(pb.Stage_Running, stCfgs), IsNil)                      // start running validator task with no error
+	t.validatorStageMatch(c, s, subtaskCfg.Name, subtaskCfg.SourceID, pb.Stage_Running)     // stage not changed
```

Comment:
why call it twice?

Reply:
To check the validator won't throw error even though the task is already started/stopped but merely log an info.